src/components/CodeMirrorEditor.jsx

```diff
@@ -0,0 +1,90 @@
+import CodeMirror from 'codemirror';
+import isNil from 'lodash-es/isNil';
+import LRU from 'lru-cache';
+import PropTypes from 'prop-types';
+import React, {useEffect, useRef} from 'react';
+
+import 'codemirror/mode/htmlmixed/htmlmixed';
+import 'codemirror/mode/css/css';
+import 'codemirror/mode/javascript/javascript';
+
+const CODEMIRROR_MODES_MAP = {
+  html: 'htmlmixed',
+  css: 'css',
+  javascript: 'javascript',
+};
+
+export default function CodeMirrorEditor({
+  language,
+  projectKey,
+  source,
+  onInput,
+}) {
+  const containerRef = useRef();
+  const editorRef = useRef();
+  const docsCacheRef = useRef(new LRU(3));
+  const mode = CODEMIRROR_MODES_MAP[language];
+
+  useEffect(() => {
+    const {current: container} = containerRef;
```

Comment:
I would normally avoid stylistic comments, but this one is bothering me. Why destructure when the key isn't used? I prefer `const container = containerRef.current;` and think it would simplify all the effects code

Reply:
Quite right! I ended up rewriting most of these already, but will seek out any lingerers. I also [added an entry to the contributor guide](https://github.com/popcodeorg/popcode/wiki/Contributor-Guide#destructuring-should-reduce-repetition-without-reducing-clarity) proposing some guidelines for when destructuring is useful.